src/DotVVM.Framework/Compilation/Binding/ExpressionHelper.cs

```diff
@@ -159,8 +160,12 @@ public static Expression CallMethod(Type target, BindingFlags flags, string name
         private static MethodRecognitionResult FindValidMethodOveloads(Type type, string name, BindingFlags flags, Type[] typeArguments, Expression[] arguments, IDictionary<string, Expression> namedArgs)
         {
             var methods = FindValidMethodOveloads(type.GetAllMembers(flags).OfType<MethodInfo>().Where(m => m.Name == name), typeArguments, arguments, namedArgs);
+            if (methods.Count() > 1)
+                // TODO: this behavior is not completed. Implement the same behavior as in roslyn.
+                throw new InvalidOperationException($"Found ambiguous overloads of method '{name}'.");
```

Comment:
We do not have to implement all the rules C# uses to distinguish the operators, but this check breaks our existing overload resolution logic - we simply prefer less automatic conversions and then less implicit type arguments. Please add unit tests for these cases

* Calling `A(1)` will prefer `void A(int)` over `void A(object)`
* Calling `A(1)` will prefer `void A(int)` over `void A<T>(T)`
* Calling `A(1)` will prefer `void A<T>(T)` over `void A(object)`

Note that this most likely works now and this line "breaks" it. Instead of checking that there is only one matching method, we should check that there is only one with the max score that we use for sorting them.

Reply:
I have added tests for the cases you have mentioned. Still, I am not a fan of just picking of method that is less automatically converted or just converted. I can't get off the feeling that someone will implement ambiguous methods and this logic will pick something else then C# would.